### PR TITLE
Completed homework10 (working with files)

### PR DESCRIPTION
**Pull request by Shammad-Kosayev**
code has been reworked to work with files

### CODE REVIEW

**Review by ayten-github — APPROVED** [review]

File: src/main/java/az/edu/strangers/controller/FamilyController.java [hunk 1]
```diff
@@ -20,70 +17,109 @@ public FamilyController(FamilyService familyService) {
...
+            @NotNull final String feminineName
+    ) {
+        if (masculineName.isEmpty() || feminineName.isEmpty())
+            throw new IllegalArgumentException("Both masculine and feminine names must be provided.");
 
         return familyService.bornChild(family, masculineName, feminineName);
     }
 
-    public Optional<Family> adoptChild(final Family family, final Human child) {
+    public Family adoptChild(@NotNull final Family family, @NotNull final Human child) {
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/controller/FamilyController.java:96]: if this annotations works? Did you test it?

File: src/main/java/az/edu/strangers/dao/CollectionFamilyDao.java [hunk 2]
```diff
@@ -2,43 +2,54 @@
...
-    public Family saveFamily(Family familyEntity) {
-        if (!FAMILY_LIST.contains(familyEntity)) {
-            FAMILY_LIST.add(familyEntity);
-        } else {
-            System.out.println("Family is already contains!");
+    public void saveFamily(Family family) {
+        for (int i = 0; i < families.size(); i++) {
+            Family existingFamily = families.get(i);
+
+            if (existingFamily.getFather().equals(family.getFather())){
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/dao/CollectionFamilyDao.java:41]: ?

File: src/main/java/az/edu/strangers/dao/FileFamilyDao.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
+package az.edu.strangers.dao;
+
+import az.edu.strangers.entity.Family;
+
+import java.io.*;
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/dao/FileFamilyDao.java:5]: avoid wildcard import

File: src/main/java/az/edu/strangers/dao/FileFamilyDao.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    private final String FILE_PATH;
+    private List<Family> families = new ArrayList<>();
+
+    public FileFamilyDao(String filePath) {
+        this.FILE_PATH = filePath;
+    }
+
+    @Override
+    public List<Family> getAllFamilies() {
+        return new ArrayList<>(families);  // Return a copy of the families list
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/dao/FileFamilyDao.java:21]: no need comment

File: src/main/java/az/edu/strangers/dao/FileFamilyDao.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    public Optional<Family> getFamilyByIndex(int index) {
+        return (index >= 0 && index < families.size()) ? Optional.of(families.get(index)) : Optional.empty();
+    }
+
+    @Override
+    public void saveFamily(Family family) {
+        for (int i = 0; i < families.size(); i++) {
+            Family existingFamily = families.get(i);
+
+            if (existingFamily.getFather().equals(family.getFather())) {
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/dao/FileFamilyDao.java:34]: ?

File: src/main/java/az/edu/strangers/dao/FileFamilyDao.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        saveToFile();
+    }
+
+    @Override
+    public boolean deleteFamily(int index) {
+        boolean result = false;
+        if (index >= 0 && index < families.size() && !families.isEmpty()) {
+            Family removed = families.remove(index);
+            result = removed != null;
+            saveToFile();
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/dao/FileFamilyDao.java:52]: why save here if you will save it again at the end of the method

File: src/main/java/az/edu/strangers/dao/FileFamilyDao.java [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+        saveToFile();
+        return result;
+    }
+
+    @Override
+    public boolean deleteFamily(Family family) {
+        boolean removed = families.remove(family);
+
+        if (removed) saveToFile();
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/dao/FileFamilyDao.java:63]: why save here if you will save it again at the end of the method

File: src/main/java/az/edu/strangers/dao/FileFamilyDao.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+        saveToFile();
+        return removed;
+    }
+
+    public void loadData(List<Family> families) {
+        this.families = new ArrayList<>(families);
+        saveToFile();  // Save the loaded data to the file system
```
Comment (sabuhihuseynov) [inline src/main/java/az/edu/strangers/dao/FileFamilyDao.java:71]: no need comment

**Review by sabuhihuseynov — COMMENTED** [review]
pls look

**Pull request merged**